File: plot_summary.m
function plot_summary(p,ref)

    spike_times=ref.sim.spike_times;
    R=ref.model.R;
    Wexc=ref.model.Wexc;
    RWinh=ref.model.RWinh; % R*Winh
    simMEP=ref.sim.simMEP2;


    figure;
    t=tiledlayout(1,4,'TileSpacing','tight','Padding','compact');
    width=15; 
    height=4.5;
    set(gcf,'units','centimeters','position',[2 2 width height])

    nexttile(1);
    space=(max(ref.y0(:))-min(ref.y0(:)))/2;
    h1=plot(ref.t0,ref.y0+repmat((1:size(ref.y0,2))*space,[size(ref.y0,1),1]),'k','linewidth',1.5);hold on;
    h2=plot(ref.t0,simMEP+repmat((1:size(ref.y0,2))*space,[size(ref.y0,1),1]),'r','linewidth',1);
    R2=1-sumsqr(ref.y0(:)-simMEP(:))/sumsqr(ref.y0(:)-mean(ref.y0(:)));
    title(sprintf('MEP (R^2= %.2g)',R2))
    set(gca,'YColor','none','box','off','tickDir','out','ytick',[])
    xlabel('Time (ms)','fontsize',10,'FontName', 'calibri')
    xlim([15 50]);ylim([0 space*length(ref.intensities)+max(ref.y0(:))])
    for i=1:length(ref.intensities)
        text(12,i*space,sprintf('%g%%',ref.intensities(i)),'fontsize',6,'FontName', 'calibri')
    end
    ylimit=ylim;
    peak=max(ref.y0(:));
    if peak>1
        plot([1;1]*47,[-1 0]+ylimit(2),'k','linewidth',1.5);
        text(46,-0.5+ylimit(2),'1 mV','fontsize',6,'FontName', 'calibri','horizontalalignment','right')
    else
        plot([1;1]*47,[-round(peak/2,1) 0]+ylimit(2),'k','linewidth',1.5);
        text(46,-round(peak/2,1)/2+ylimit(2),sprintf('%g mV',round(peak/2,1)),'fontsize',6,'FontName', 'calibri','horizontalalignment','right')
    end

    % IO curve
    nexttile(2);
    [IO,simIO,myfit1,myfit2]=get_iocurve(simMEP,ref); 
    x=linspace(IO(1,1),IO(end,1),100);
    plot(x,myfit1(x),'k','linewidth',1); hold on;
    x=linspace(simIO(1,1),simIO(end,1),100);
    plot(x,myfit2(x),'r','linewidth',1);
    h1=scatter(IO(:,1),IO(:,2),15,'ko');
    errorbar(IO(:,1),IO(:,2),IO(:,3),'k','LineStyle','none')
    h2=scatter(simIO(:,1),simIO(:,2),15,'ro');    
    xlim([27 58]);box off;
    xlabel('TMS intensity (%MSO)','fontsize',10,'FontName', 'calibri');
    ylabel('Amplitude (mV)','fontsize',8,'FontName', 'calibri');
    title('IO curve');
    if ismember(ref.subj,[1,2,4,7,9])
        yticks=unique([get(gca,'ytick'),0.5]);
        set(gca,'ytick',yticks)
    end
    
    nexttile(3);
    if ref.model.withRC 
        h1=scatter(1:100,Wexc,10,'b.');hold on;    
        h2=scatter(1:100,RWinh,10,'m.');
    end
    h3=scatter(1:100,R,10,'k.'); hold on;
    scatter([1 10 20 60 100],R([1 10 20 60 100]),20,'ko'); 

    title('Model param.');
    ylimit=ylim;
    text(15,ylimit(2)*0.85,['R: ',sprintf('[%g, %g]',round([R(1),R(100)],1))],...
         'color','k','fontsize',7,'FontName', 'calibri','BackgroundColor','none')
    if ref.model.withRC
        text(15,ylimit(2)*0.75,['Wexc: ',sprintf('[%g, %g]',round([Wexc(1),Wexc(100)],1))],...
             'color','b','fontsize',7,'FontName', 'calibri','BackgroundColor','none')
        text(15,ylimit(2)*0.65,['Winh*R: ',sprintf('[%g, %g]',round([RWinh(1),RWinh(100)],1))],...
             'color','m','fontsize',7,'FontName', 'calibri','BackgroundColor','none')
        text(30,ylimit(2)*0.55,sprintf('RCth= %g',round(p(10),1)),...
             'color','k','fontsize',7,'FontName', 'calibri','BackgroundColor','none')
    end
 
    text(30,ylimit(2)*0.45,sprintf('dAxon= %g',round(ref.model.axonalDelay,1)),...
        'color','k','fontsize',7,'FontName', 'calibri','BackgroundColor','none')
    text(30,ylimit(2)*0.35,sprintf('Tmu= %g',round(ref.model.Tmu,1)),...
        'color','k','fontsize',7,'FontName', 'calibri','BackgroundColor','none')
    text(15,ylimit(2)*0.95,sprintf('AMPAw= %g',round(ref.model.AMPAweight,1)),...
        'color','k','fontsize',7,'FontName', 'calibri','BackgroundColor','none')

    xlabel('Motor neuron','fontsize',10,'FontName', 'calibri');
    set(gca,'xtick',[1,50,100],'xticklabel',[1,50,100]);
     

    nexttile(4);
    switch 2
        case 1
            plot(reshape(spike_times+ref.model.axonalDelay,100*ref.model.maxES,[]),repmat(1:100,[1,ref.model.maxES]),'.');hold on;
            ylim([0 100]);xlim([15 50]);
            ylabel('Motor unit');set(gca,'ytick',[1,50,100],'yticklabel',[1,50,100])
        case 2 
            tmp=permute(spike_times+ref.model.axonalDelay,[1 3 2]);
            tmp=reshape(tmp,100*length(ref.intensities),ref.model.maxES); % [100*Int x maxEs]
            scatter(tmp,1:(100*length(ref.intensities)),5,'k.');hold on;
            ylim([0 100*length(ref.intensities)]);xlim([15 50]);
            plot([17 50],[1;1]*100*(1:length(ref.intensities)-1),'k-')
            for i=1:length(ref.intensities)
                text(12,i*100-50,sprintf('%g%%',ref.intensities(i)),'fontsize',6,'FontName', 'calibri')
            end
            set(gca,'YColor','none','box','off','tickDir','out','ytick',[])
    end
    plot([1;1]*47,[1,100],'k','linewidth',1.5); 
    text(45,50,'100 MUs','fontsize',8,'FontName', 'calibri','horizontalalignment','right')    
    xlabel('Time (ms)','fontsize',10,'FontName', 'calibri')
    title('MU trigger time');
    %axis square
    if isfield(ref,'figname')
        root=fileparts(mfilename("fullpath"));
        saveas(gcf,[root,filesep,ref.figname])
    end
end

%==========================================================================
function [IO,simIO,myfit1,myfit2]=get_iocurve(simMEP,ref)

    % reload single-trial MEP for std of IO curve
    [~,~,~,~,~,y0all] = load_MEP(ref.subj,ref.intensity_idx,[20,50],0);
    %-------------------------------------------
    IO=zeros(length(ref.intensity_idx),3); % [Int,amplitude,std] 
    IOall=zeros(length(ref.intensity_idx),2,size(y0all,3)); % 15 trials
    simIO=zeros(length(ref.intensity_idx),2);
    for i=1:length(ref.intensity_idx)
        tidx1=find(ref.t0>=24 & ref.t0<=28);
        [peakv1,peaki1]=max(ref.y0(tidx1,i)); %t=23.9~27.9 ms
        tidx2=find(ref.t0>=ref.t0(tidx1(peaki1)) & ref.t0<=ref.t0(tidx1(peaki1))+6); % trough within 6 ms
        [peakv2,~]=min(ref.y0(tidx2,i)); %t=23.9~27.9 ms
        IO(i,1)=ref.intensities(i);
        IO(i,2)=peakv1-peakv2;
        tmp=zeros(1,size(y0all,3)); % 15 trials
        for j=1:size(y0all,3) % 15 trials
           peakv1=max(y0all(i,tidx1,j)); %t=23.9~27.9 ms
           peakv2=min(y0all(i,tidx2,j)); % trough within 6 ms
           tmp(j)=peakv1-peakv2;
        end
        IO(i,3)=std(tmp)/sqrt(length(tmp)); % std error of MEP amplitude
        %----
        tidx=find(ref.t0>=24 & ref.t0<=28);
        [peakv1,peaki1]=max(simMEP(tidx,i)); %t=23.9~27.9 ms
        tidx=find(ref.t0>=ref.t0(tidx(peaki1)) & ref.t0<=ref.t0(tidx(peaki1))+6); % trough within 6 ms
        [peakv2,~]=min(simMEP(tidx,i)); %t=23.9~27.9 ms
        simIO(i,1)=ref.intensities(i);
        simIO(i,2)=peakv1-peakv2;
    end
    myfittype = fittype('a/(1+exp(r*(x0-x)))',...
    'dependent',{'y'},'independent',{'x'},...
    'coefficients',{'a','r','x0'});
    switch ref.subj
     case {1,2,3,4,9,10}
         [myfit1,gof,output] = fit(IO(:,1),IO(:,2),myfittype,'StartPoint',[1.4, 10, 40]);
         [myfit2,gof,output] = fit(simIO(:,1),simIO(:,2),myfittype,'StartPoint',[1.4, 10, 40]);
     case {5,7,8}
         [myfit1,gof,output] = fit(IO(:,1),IO(:,2),myfittype,'StartPoint',[1.4, 5, 50]);
         [myfit2,gof,output] = fit(simIO(:,1),simIO(:,2),myfittype,'StartPoint',[1.4, 5, 50]);
     case 6
         [myfit1,gof,output] = fit(IO(:,1),IO(:,2),myfittype,'StartPoint',[1.4, 2, 60]);
         [myfit2,gof,output] = fit(simIO(:,1),simIO(:,2),myfittype,'StartPoint',[1.4, 2, 60]);
    end
end

File: plot_summary.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec
#from scipy.interpolate import interp1d
from scipy.optimize import curve_fit
from sigmoid import sigmoid
from load_MEP import load_MEP

from scipy.io import loadmat

def plot_summary(p, ref):
    # Extracting variables from the nested ref dictionary using tuple keys
    spike_times = ref['sim']['spike_times']
    R = ref['model']['R']
    Wexc = ref['model']['Wexc']
    RWinh = ref['model']['RWinh']  # R*Winh
    simMEP = ref['sim']['simMEP2']
    
    # Simple keys
    y0 = ref['y0']
    t0 = ref['t0']
    intensities = ref['intensities']

    fig = plt.figure(figsize=(20/2.54, 6/2.54)) # Convert cm to inches
    gs = GridSpec(1, 4, figure=fig, wspace=0.55, left=0.05, right=0.95, top=0.85, bottom=0.2)

    # --- Nexttile(1): MEP comparison ---
    ax1 = fig.add_subplot(gs[0, 0])
    space = (np.max(y0) - np.min(y0)) / 2
    
    # MATLAB: repmat((1:size(ref.y0,2))*space,[size(ref.y0,1),1])
    # y0 is assumed [time x intensities]
    offsets = np.arange(1, y0.shape[1] + 1) * space
    
    h1 = ax1.plot(t0, y0 + offsets, 'k', linewidth=1.5)
    h2 = ax1.plot(t0, simMEP + offsets, 'r', linewidth=1)
    
    # R^2 calculation
    R2 = 1 - np.sum((y0.flatten() - simMEP.flatten())**2) / np.sum((y0.flatten() - np.mean(y0.flatten()))**2)
    ax1.set_title(f'MEP (R^2= {R2:.2g})', fontsize=9)
    ax1.spines['left'].set_visible(False)
    ax1.spines['right'].set_visible(False)
    ax1.spines['top'].set_visible(False)
    ax1.set_yticks([])
    ax1.set_xlabel('Time (ms)', fontsize=10)
    ax1.set_xlim([15, 50])
    ax1.set_ylim([0, space * len(intensities) + np.max(y0)])
    
    for i in range(len(intensities)):
        ax1.text(12, (i + 1) * space, f'{intensities[i]}%', fontsize=6)
    
    ylimit = ax1.get_ylim()
    peak = np.max(y0)
    if peak > 1:
        ax1.plot([47, 47], [ylimit[1] - 1, ylimit[1]], 'k', linewidth=1.5)
        ax1.text(46, ylimit[1] - 0.5, '1 mV', fontsize=6, ha='right')
    else:
        v_scale = np.round(peak / 2, 1)
        ax1.plot([47, 47], [ylimit[1] - v_scale, ylimit[1]], 'k', linewidth=1.5)
        ax1.text(46, ylimit[1] - v_scale / 2, f'{v_scale} mV', fontsize=6, ha='right')

    # --- Nexttile(2): IO Curve ---
    ax2 = fig.add_subplot(gs[0, 1])
    # Assume get_iocurve is defined elsewhere as per prompt
    IO, simIO, myfit1, myfit2 = get_iocurve(simMEP, ref)
    
    x1 = np.linspace(IO[0, 0], IO[-1, 0], 100)
    ax2.plot(x1, sigmoid(x1, *myfit1), 'k', linewidth=1)
  
    x2 = np.linspace(simIO[0, 0], simIO[-1, 0], 100)
    ax2.plot(x2, sigmoid(x2, *myfit2), 'r', linewidth=1)
    
    ax2.scatter(IO[:, 0], IO[:, 1], 15, edgecolors='k', facecolors='none')
    ax2.errorbar(IO[:, 0], IO[:, 1], yerr=IO[:, 2], fmt='none', ecolor='k')
    ax2.scatter(simIO[:, 0], simIO[:, 1], 15, edgecolors='r', facecolors='none')
    
    ax2.set_xlim([27, 58])
    ax2.spines['top'].set_visible(False)
    ax2.spines['right'].set_visible(False)
    ax2.set_xlabel('TMS intensity (%MSO)', fontsize=10)
    ax2.set_ylabel('Amplitude (mV)', fontsize=8)
    ax2.set_title('IO curve', fontsize=9)
    ax2.spines['top'].set_visible(False)
    ax2.spines['right'].set_visible(False)

    if ref['subj'] in [1, 2, 4, 7, 9]:
        ticks = np.unique(np.append(ax2.get_yticks(), 0.5))
        ax2.set_yticks(ticks)

    # --- Nexttile(3): Model Parameters ---
    ax3 = fig.add_subplot(gs[0, 2])
    mn_indices = np.arange(1, 101)
    
    if 'withRC' in ref['model']:
        ax3.scatter(mn_indices, Wexc, 10, c='b', marker='.', alpha=0.5)
        ax3.scatter(mn_indices, RWinh, 10, c='m', marker='.', alpha=0.5)
    
    ax3.scatter(mn_indices, R, 10, c='k', marker='.')
    hi_idx = np.array([0, 9, 19, 59, 99]) # 1, 10, 20, 60, 100 in 0-based
    ax3.scatter(mn_indices[hi_idx], R[hi_idx], 20, edgecolors='k', facecolors='none')

    ax3.set_title('Model param.', fontsize=9)
    ylimit3 = ax3.get_ylim()
    
    text_params = {'fontsize': 7, 'backgroundcolor': 'none'}
    ax3.text(15, ylimit3[1] * 0.8, f'R: [{R[0]:.1f}, {R[99]:.1f}]', color='k', **text_params)
    
    if 'withRC' in ref['model']:
        ax3.text(15, ylimit3[1] * 0.7, f'Wexc: [{Wexc[0]:.1f}, {Wexc[99]:.1f}]', color='b', **text_params)
        ax3.text(15, ylimit3[1] * 0.6, f'Winh*R: [{RWinh[0]:.1f}, {RWinh[99]:.1f}]', color='m', **text_params)
        ax3.text(30, ylimit3[1] * 0.5, f'RCth= {np.round(p[9], 1)}', color='k', **text_params)

    ax3.text(30, ylimit3[1] * 0.4, f"dAxon= {np.round(ref['model']['axonalDelay'] , 1)}", color='k', **text_params)
    ax3.text(30, ylimit3[1] * 0.3, f"Tmu= {np.round(ref['model']['Tmu'] , 1)}", color='k', **text_params)
    ax3.text(15, ylimit3[1] * 0.9, f"AMPAw= {np.round(ref['model']['AMPAweight'] , 1)}", color='k', **text_params)

    ax3.set_xlabel('Motor neuron', fontsize=10)
    ax3.set_xticks([1, 50, 100])
    ax3.set_xlim([1, 100])
    ax3.spines['top'].set_visible(False)
    ax3.spines['right'].set_visible(False)

    # --- Nexttile(4): MU Trigger Time ---
    ax4 = fig.add_subplot(gs[0, 3])
    axonal_delay = ref['model']['axonalDelay'] 
    maxES = ref['model']['maxES']
    
    # Using 'case 2' logic from MATLAB
    tmp = np.transpose(spike_times + axonal_delay, (0, 2, 1)) # [100 x intensities x maxES]
    tmp_flat = np.reshape(tmp, (100 * len(intensities), maxES), order='F')

    y_coords = np.arange(1, (100 * len(intensities)) + 1)
    for col in range(maxES):
        ax4.scatter(tmp_flat[:, col], y_coords, 0.2, c='k', marker='.')
        
    ax4.set_ylim([0, 100 * len(intensities)])
    ax4.set_xlim([15, 50])
    
    # Intensity boundary lines
    for i in range(1, len(intensities)):
        ax4.plot([17, 50], [i * 100, i * 100], 'k-', linewidth=0.5)
        
    for i in range(len(intensities)):
        ax4.text(12, (i + 0.7) * 100 - 50, f'{intensities[i]}%', fontsize=6)
        
    ax4.spines['left'].set_visible(False)
    ax4.set_yticks([])
    ax4.plot([47, 47], [1, 100], 'k', linewidth=1.5)
    ax4.text(45, 10, '100 MUs', fontsize=8, ha='right')
    ax4.set_xlabel('Time (ms)', fontsize=10)
    ax4.set_title('MU trigger time', fontsize=9)
    ax4.spines['top'].set_visible(False)
    ax4.spines['right'].set_visible(False)

    plt.show()

def get_iocurve(simMEP, ref):
    # reload single-trial MEP for std of IO curve
    _, _, _, _, _, y0all = load_MEP(ref["subj"], ref["intensity_idx"], [20, 50], 0)

    IO = np.zeros((len(ref["intensity_idx"]), 3))  # [Int, amplitude, std]
    simIO = np.zeros((len(ref["intensity_idx"]), 2))

    for i in range(len(ref["intensity_idx"])):
        tidx1 = np.where((ref["t0"] >= 24) & (ref["t0"] <= 28))[0]
        peakv1 = np.max(ref["y0"][tidx1, i])
        peaki1 = np.argmax(ref["y0"][tidx1, i])
        tidx2 = np.where((ref["t0"] >= ref["t0"][tidx1[peaki1]]) & (ref["t0"] <= ref["t0"][tidx1[peaki1]] + 6))[0]
        peakv2 = np.min(ref["y0"][tidx2, i])
        IO[i, 0] = ref["intensities"][i]
        IO[i, 1] = peakv1 - peakv2

        tmp = np.zeros(y0all.shape[2])  # 15 trials
        for j in range(y0all.shape[2]):
            peakv1 = np.max(y0all[i, tidx1, j])
            peakv2 = np.min(y0all[i, tidx2, j])
            tmp[j] = peakv1 - peakv2
        IO[i, 2] = np.std(tmp, ddof=1) / np.sqrt(len(tmp))  # std error of MEP amplitude

        tidx = np.where((ref["t0"] >= 24) & (ref["t0"] <= 28))[0]
        peakv1 = np.max(simMEP[tidx, i])
        peaki1 = np.argmax(simMEP[tidx, i])
        tidx = np.where((ref["t0"] >= ref["t0"][tidx[peaki1]]) & (ref["t0"] <= ref["t0"][tidx[peaki1]] + 6))[0]
        peakv2 = np.min(simMEP[tidx, i])
        simIO[i, 0] = ref["intensities"][i]
        simIO[i, 1] = peakv1 - peakv2

    if ref["subj"] in {1, 2, 3, 4, 9, 10}:
        p0 = [40, 1.4, 10]
    elif ref["subj"] in {5, 7, 8}:
        p0 = [50, 1.4, 5]
    elif ref["subj"] == 6:
        p0 = [60, 1.4, 2]
    else:
        p0 = [40, 1.4, 10]  # default fallback

    popt1, pcov1 = curve_fit(sigmoid, IO[:, 0], IO[:, 1], p0=p0)
    popt2, pcov2 = curve_fit(sigmoid, simIO[:, 0], simIO[:, 1], p0=p0)

    myfit1 = sigmoid(IO[:, 0], *popt1)
    myfit2 = sigmoid(simIO[:, 0], *popt2)

    return IO, simIO, popt1, popt2